pkg/collector/smbclient/smbclient.go

```diff
@@ -0,0 +1,502 @@
+
+	WriteRequestsQueued                        *prometheus.Desc
+	WriteBytesTotal                            *prometheus.Desc
+	WritesTotal                                *prometheus.Desc
+	WriteBytesTransmittedViaSMBDirectTotal     *prometheus.Desc
+	WriteRequestsTransmittedViaSMBDirectTotal  *prometheus.Desc
+	TurboIOWritesTotal                         *prometheus.Desc
+	WriteSecsTotal                             *prometheus.Desc
+
+	RequestsQueued                             *prometheus.Desc
+	RequestSecs                                *prometheus.Desc
+	CreditStallsTotal                          *prometheus.Desc
+	CurrentDataQueued                          *prometheus.Desc
+	DataBytesTotal                             *prometheus.Desc
+	DataRequestsTotal                          *prometheus.Desc
+	MetadataRequestsTotal                     *prometheus.Desc
+
+	enabledCollectors []string
+}
+
+// All available collector functions
+var smbclientAllCollectorNames = []string{
+	"ClientShares",
+}
+
+func New(logger log.Logger, config *Config) types.Collector {
+	if config == nil {
+		config = &ConfigDefaults
+	}
+
+	smbclientListAllCollectors := false
+	c := &collector{
+		smbclientCollectorsEnabled: &config.CollectorsEnabled,
+		smbclientListAllCollectors: &smbclientListAllCollectors,
+	}
+	c.SetLogger(logger)
+	return c
+}
+
+func NewWithFlags(app *kingpin.Application) types.Collector {
+	return &collector{
+		smbclientListAllCollectors: app.Flag(
+			FlagSmbClientListAllCollectors,
+			"List the collectors along with their perflib object name/ids",
+		).Bool(),
+
+		smbclientCollectorsEnabled: app.Flag(
+			FlagSmbClientCollectorsEnabled,
+			"Comma-separated list of collectors to use. Defaults to all, if not specified.",
+		).Default(ConfigDefaults.CollectorsEnabled).String(),
+	}
+}
+
+func (c *collector) GetName() string {
+	return Name
+}
+
+func (c *collector) SetLogger(logger log.Logger) {
+	c.logger = log.With(logger, "collector", Name)
+}
+
+func (c *collector) GetPerfCounter() ([]string, error) {
+	return []string{
+		"SMB Client Shares",
+	}, nil
+}
+
+func (c *collector) Build() error {
+	// desc creates a new prometheus description
+	desc := func(metricName string, description string, labels []string) *prometheus.Desc {
+		return prometheus.NewDesc(
+			prometheus.BuildFQName(types.Namespace, "smbclient", metricName),
+			description,
+			labels,
+			nil,
+		)
+	}
+
+	// These appear to be equivalent to bytes read/write total so
+	// redundant and confusing to include.
+	//
+	// c.AvgBytesPerRead = desc(
+	// 	"bytes_per_read",
+	// 	"The average number of bytes per read request.",
+	//   []string{"server", "share"},
+	// 	)
+	// c.AvgBytesPerWrite = desc("bytes_per_write",
+	// 	"The average number of bytes per write request.",
+	//   []string{"server", "share"},)
+	// c.AvgDataBytesPerRequest = desc("data_bytes_per_request",
+	// 	"The average number of bytes per read or write request.",
+	//   []string{"server", "share"},)
```

Comment:
remove comment out code or include it

Reply:
Hi I'm Bob. I work with Rob on windows exporter. 

That's my code, I will be happy to remove the commented out code. I would like to include any counter considered useful, but my experience with SMB is limited to the workloads we are running now (simple read and write). I was curious to know if there is anyway to find out more about the perflib or counter definition. I read the comments in perflib package. The closest I came was using `Get-Counter` commands but didn't see any help for a specific counter, just the counter set. Is there a way to get more information about the definition of a counter? Otherwise, I guess it's just a best-effort process. Thanks.

Reply:
Site Note: In my eye, comment code is dead code and mostly will not survive any refactoring.


About the Perf Counter questions:

The GUI Tool "Performance Monitor" is a builtin tool (perfmon.msc) to obverse and see the existing Performance Counters in a System. 

See also: https://www.progress.com/blogs/windows-performance-counters-getting-started